### PR TITLE
Bump golangci-lint to be compatible with go 1.20

### PR DESCRIPTION
**Pull request by hpidcock**
Also fixes a few bugs with exported loop variables, the rest are to silence the new linter, exportloopref.

Fixes a few errors picked up by ineffassign.

## QA steps

Run linters, unit tests, check the changes made.

## Documentation changes

N/A

## Bug reference

N/A

### CODE REVIEW

File: .github/golangci-lint.config.yaml [hunk 1]
```diff
@@ -32,8 +32,7 @@ linters:
     - ineffassign
     - misspell
     - unconvert
-    - deadcode
```
Comment (hpidcock) [inline .github/golangci-lint.config.yaml:35]: these are deprecated

File: .github/workflows/static-analysis.yml [hunk 2]
```diff
@@ -48,7 +48,7 @@ jobs:
         echo "GOPATH=$(go env GOPATH)" >> $GITHUB_ENV
         echo "$(go env GOPATH)/bin" >> $GITHUB_PATH
 
-        curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.46.2
+        curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.52.2
```
Comment (hpidcock) [inline .github/workflows/static-analysis.yml:51]: the old version would panic with go 1.20 installed locally

File: cmd/juju/storage/add.go [hunk 3]
```diff
@@ -202,14 +202,15 @@ type StorageAddAPI interface {
...
+	for one, sc := range c.storageCons {
+		cons := sc
 		all = append(all, params.StorageAddParams{
 			UnitTag:     c.unitTag.String(),
 			StorageName: one,
 			Constraints: params.StorageConstraints{
-				cons.Pool,
-				&cons.Size,
-				&cons.Count,
+				Pool:  cons.Pool,
```
Comment (hpidcock) [inline cmd/juju/storage/add.go:211]: Possible breaking/fixing change.

File: cmd/juju/action/cancel.go [hunk 4]
```diff
@@ -113,7 +113,8 @@ func (c *cancelCommand) Run(ctx *cmd.Context) error {
 	var failedCancels []unCanceledAction
 	var canceledActions []actionapi.ActionResult
 
-	for i, result := range actions {
+	for i, v := range actions {
+		result := v
```
Comment (hpidcock) [inline cmd/juju/action/cancel.go:117]: Possible breaking/fixing change.

File: apiserver/facades/client/modelmanager/modelmanager.go [hunk 5]
```diff
@@ -903,14 +903,15 @@ func (m *ModelManagerAPI) ListModels(user params.Entity) (params.UserModelList,
...
+		lastConnection := mi.LastConnection
 		result.UserModels = append(result.UserModels, params.UserModel{
 			Model: params.Model{
 				Name:     mi.Name,
 				UUID:     mi.UUID,
 				Type:     string(mi.Type),
 				OwnerTag: ownerTag.String(),
 			},
-			LastConnection: &mi.LastConnection,
+			LastConnection: &lastConnection,
```
Comment (hpidcock) [inline apiserver/facades/client/modelmanager/modelmanager.go:914]: Possible breaking/fixing change.

File: environs/manual/sshprovisioner/sshprovisioner.go [hunk 6]
```diff
@@ -127,6 +127,9 @@ func detectSeriesAndHardwareCharacteristics(host string) (hc instance.HardwareCh
 	memkB := strings.Fields(lines[2])[1] // "MemTotal: NNN kB"
 	hc.Mem = new(uint64)
 	*hc.Mem, err = strconv.ParseUint(memkB, 10, 0)
+	if err != nil {
+		return hc, "", errors.Annotatef(err, "parsing %q", lines[2])
```
Comment (hpidcock) [inline environs/manual/sshprovisioner/sshprovisioner.go:131]: Possible breaking/fixing change.

File: state/migration_export.go [hunk 7]
```diff
@@ -1830,7 +1830,8 @@ func (e *exporter) readAllMeterStatus() (map[string]*meterStatusDoc, error) {
 	}
 	e.logger.Debugf("found %d meter status docs", len(docs))
 	result := make(map[string]*meterStatusDoc)
-	for _, doc := range docs {
+	for _, v := range docs {
+		doc := v
```
Comment (hpidcock) [inline state/migration_export.go:1834]: Possible breaking/fixing change.

File: state/migration_export.go [hunk 8]
```diff
@@ -1864,7 +1865,8 @@ func (e *exporter) readAllCloudServices() (map[string]*cloudServiceDoc, error) {
 	}
 	e.logger.Debugf("found %d cloud service docs", len(docs))
 	result := make(map[string]*cloudServiceDoc)
-	for _, doc := range docs {
+	for _, v := range docs {
```
Comment (hpidcock) [inline state/migration_export.go:1868]: Possible breaking/fixing change.

File: state/migration_export.go [hunk 9]
```diff
@@ -1888,7 +1890,8 @@ func (e *exporter) readAllCloudContainers() (map[string]*cloudContainerDoc, erro
 	}
 	e.logger.Debugf("found %d cloud container docs", len(docs))
 	result := make(map[string]*cloudContainerDoc)
-	for _, doc := range docs {
+	for _, v := range docs {
+		doc := v
```
Comment (hpidcock) [inline state/migration_export.go:1894]: Possible breaking/fixing change.

File: worker/uniter/charm/bundles.go [hunk 10]
```diff
@@ -69,6 +69,9 @@ func (d *BundlesDir) download(info BundleInfo, target string, abort <-chan struc
 		return errors.Annotate(err, "could not parse charm URL")
 	}
 	expectedSha256, err := info.ArchiveSha256()
+	if err != nil {
+		return errors.Annotatef(err, "failed to get archive sha256 for charm %q", info.String())
```
Comment (hpidcock) [inline worker/uniter/charm/bundles.go:73]: Possible breaking/fixing change.

File: worker/uniter/runner/jujuc/storage-add.go [hunk 11]
```diff
@@ -47,17 +47,18 @@ func (s *StorageAddCommand) Init(args []string) error {
...
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	s.all = make(map[string]params.StorageConstraints, len(cons))
-	for k, v := range cons {
-		if v != (storage.Constraints{Count: v.Count}) {
+	s.all = make(map[string]params.StorageConstraints, len(constraintsMap))
+	for k, v := range constraintsMap {
+		cons := v
```
Comment (hpidcock) [inline worker/uniter/runner/jujuc/storage-add.go:57]: Possible breaking/fixing change.

File: .github/golangci-lint.config.yaml (hunk 1 above)
Comment (wallyworld) [reply .github/golangci-lint.config.yaml:35]: is there an equivalent newer check?

**Review by wallyworld — APPROVED** [review]
Thanks for fixing the for loop issues.
How have we not seen more bugs due to these.

File: .github/golangci-lint.config.yaml (hunk 1 above)
Comment (hpidcock) [reply .github/golangci-lint.config.yaml:35]: deadcode => unused
varcheck => unused

Comment (hpidcock) [reply .github/golangci-lint.config.yaml:35]: will include in a followup pr.

**Pull request merged**